File: backend/app/core/response_cache.py
"""Server-side response caching for high-traffic read endpoints.

Uses Redis (with local in-memory fallback) to cache API responses.
Reduces database load for leaderboard, dashboard, analytics, and passage data.

Cache keys are auto-invalidated when related data changes.
"""
import time
import json
import hashlib
from typing import Optional, Any, Callable, Awaitable
from collections import OrderedDict

from app.core.cache import cache


class ResponseCache:
    TTL_DASHBOARD = 180
    TTL_LEADERBOARD = 120
    TTL_ANALYTICS = 300
    TTL_PASSAGE_LIST = 300
    TTL_PASSAGE_RANDOM = 600
    TTL_USER_PROFILE = 60

    _local_cache: OrderedDict[str, tuple[float, Any]] = OrderedDict()
    _LOCAL_MAX = 500

    async def get(self, key: str) -> Optional[Any]:
        data = await cache.get(self._ck(key))
        if data is not None:
            return data
        return self._local_get(key)

    async def set(self, key: str, value: Any, ttl: int = 60) -> None:
        await cache.set(self._ck(key), value, ttl=ttl)
        self._local_set(key, value, ttl)

    async def invalidate(self, key: str) -> None:
        await cache.delete(self._ck(key))
        self._local_cache.pop(key, None)

    async def invalidate_prefix(self, prefix: str) -> None:
        client = await cache.get_client()
        if client:
            cursor = 0
            pattern = self._ck(f"{prefix}*")
            while True:
                cursor, keys = await client.scan(cursor=cursor, match=pattern, count=100)
                if keys:
                    await client.delete(*keys)
                if cursor == 0:
                    break
        self._local_cache = OrderedDict(
            (k, v) for k, v in self._local_cache.items()
            if not k.startswith(prefix)
        )

    async def get_or_compute(
        self,
        key: str,
        ttl: int,
        factory: Callable[[], Awaitable[Any]],
    ) -> Any:
        cached = await self.get(key)
        if cached is not None:
            return cached
        value = await factory()
        await self.set(key, value, ttl)
        return value

    def cache_key(self, *parts: str) -> str:
        return hashlib.md5(":".join(parts).encode()).hexdigest()

    def _ck(self, key: str) -> str:
        return f"rc:{key}"

    def _local_get(self, key: str) -> Optional[Any]:
        if key in self._local_cache:
            expiry, value = self._local_cache[key]
            if time.time() < expiry:
                return value
            del self._local_cache[key]
        return None

    def _local_set(self, key: str, value: Any, ttl: int) -> None:
        self._local_cache[key] = (time.time() + ttl, value)
        self._local_cache.move_to_end(key)
        while len(self._local_cache) > self._LOCAL_MAX:
            self._local_cache.popitem(last=False)


response_cache = ResponseCache()

File: frontend/src/server/services/response-cache.ts
import { RedisCache } from '../redis/cache';
import { redis } from '../redis/client';
import crypto from 'crypto';

class ResponseCacheService {
  private localCache: Map<string, { expiry: number; value: any }> = new Map();
  private static readonly LOCAL_MAX = 500;

  private static readonly TTL = {
    dashboard: 180,
    leaderboard: 120,
    analytics: 300,
    passageList: 300,
    passageRandom: 600,
    userProfile: 60,
  } as const;

  async get(key: string): Promise<any | null> {
    const redisData = await RedisCache.get<any>(this.cacheKey(key));
    if (redisData !== null) return redisData;
    return this.localGet(key);
  }

  async set(key: string, value: any, ttl: number = 60): Promise<void> {
    await RedisCache.set(this.cacheKey(key), value, ttl);
    this.localSet(key, value, ttl);
  }

  async invalidate(key: string): Promise<void> {
    await RedisCache.del(this.cacheKey(key));
    this.localCache.delete(key);
  }

  async invalidatePrefix(prefix: string): Promise<void> {
    let cursor = '0';
    const pattern = `rc:${prefix}*`;
    do {
      const result = await redis.scan(cursor, 'MATCH', pattern, 'COUNT', 100);
      cursor = result[0];
      const keys = result[1];
      if (keys.length > 0) {
        await redis.del(...keys);
      }
    } while (cursor !== '0');

    for (const key of this.localCache.keys()) {
      if (key.startsWith(prefix)) {
        this.localCache.delete(key);
      }
    }
  }

  async getOrCompute(
    key: string,
    ttl: number,
    factory: () => Promise<any>,
  ): Promise<any> {
    const cached = await this.get(key);
    if (cached !== null) return cached;
    const value = await factory();
    await this.set(key, value, ttl);
    return value;
  }

  makeCacheKey(...parts: string[]): string {
    return crypto.createHash('md5').update(parts.join(':')).digest('hex');
  }

  private cacheKey(key: string): string {
    return `rc:${key}`;
  }

  private localGet(key: string): any | null {
    const entry = this.localCache.get(key);
    if (!entry) return null;
    if (Date.now() < entry.expiry) return entry.value;
    this.localCache.delete(key);
    return null;
  }

  private localSet(key: string, value: any, ttl: number): void {
    this.localCache.set(key, { expiry: Date.now() + ttl * 1000, value });
    if (this.localCache.size > ResponseCacheService.LOCAL_MAX) {
      const firstKey = this.localCache.keys().next().value;
      if (firstKey) this.localCache.delete(firstKey);
    }
  }
}

export const responseCache = new ResponseCacheService();
